File: entities/player.lua
local Player = {}
Player.__index = Player

Player.sprites = nil

function Player.loadSprites()
    if Player.sprites then
        return
    end

    Player.sprites = {}
    for i = 1, 4 do
        local path = string.format("sprites/player/p_0%d.png", i)
        Player.sprites[i] = love.graphics.newImage(path)
    end
end

function Player.new(x, y)
    Player.loadSprites()

    local self = setmetatable({}, Player)
    self.x = x
    self.y = y
    self.width = 20
    self.height = 20

    self.baseSpeed = 300
    self.speed = self.baseSpeed
    -- animation
    self.currentFrame = 1
    self.animationTimer = 0
    self.animationSpeed = 0.1 -- seconds per frame

    return self
end

function Player:update(dt, gameWidth, gameHeight)
    -- update animation
    self.animationTimer = self.animationTimer + dt
    if self.animationTimer >= self.animationSpeed then
        self.animationTimer = self.animationTimer - self.animationSpeed
        self.currentFrame = self.currentFrame + 1
        if self.currentFrame > 4 then
            self.currentFrame = 1
        end
    end

    local dx = 0
    local dy = 0

    if love.keyboard.isDown('w') or love.keyboard.isDown('up') then
        dy = dy - 1
    end
    if love.keyboard.isDown('s') or love.keyboard.isDown('down') then
        dy = dy + 1
    end
    if love.keyboard.isDown('a') or love.keyboard.isDown('left') then
        dx = dx - 1
    end
    if love.keyboard.isDown('d') or love.keyboard.isDown('right') then
        dx = dx + 1
    end

    -- normalize
    if dx ~= 0 or dy ~= 0 then
        local magnitude = math.sqrt(dx * dx + dy * dy)
        dx = dx / magnitude
        dy = dy / magnitude
    end

    self.x = self.x + dx * self.speed * dt
    self.y = self.y + dy * self.speed * dt

    -- clamp 
    if self.x < 0 then
        self.x = 0
    elseif self.x + self.width > gameWidth then
        self.x = gameWidth - self.width
    end

    if self.y < 0 then
        self.y = 0
    elseif self.y + self.height > gameHeight then
        self.y = gameHeight - self.height
    end
end

function Player:draw()
    local sprite = Player.sprites[self.currentFrame]

    if sprite then
        local centerX = self.x + self.width / 2
        local centerY = self.y + self.height / 2

        local spriteWidth = sprite:getWidth()
        local spriteHeight = sprite:getHeight()
        local scaleX = self.width / spriteWidth
        local scaleY = self.height / spriteHeight
        local scale = math.min(scaleX, scaleY)

        love.graphics.setColor(1, 1, 1, 1)
        love.graphics.draw(sprite, centerX, centerY, 0, scale, scale, spriteWidth / 2, spriteHeight / 2)
    else
        love.graphics.setColor(0, 1, 0)
        love.graphics.rectangle('fill', self.x, self.y, self.width, self.height)
    end

    love.graphics.setColor(1, 1, 1)
end

function Player:getCollisionBox()
    return self.x, self.y, self.width, self.height
end

return Player

File: cpp_migration_entities/player.cpp
#include "player.h"

Player::Player(float x, float y, int w, int h, int base_speed_param)
    : generic_object(x, y, w, h, "sprites/player/")
{
    this->base_speed = static_cast<float>(base_speed_param);
    this->current_speed = this->base_speed;
}

void Player::update(float dt, int view_w, int view_h)
{
    // animation: only if we have frames
    if (!textures.empty())
    {
        animation_timer += dt;
        if (animation_timer >= animation_speed)
        {
            animation_timer = 0;
            current_frame = (current_frame + 1) % static_cast<int>(textures.size());
            sprite.setTexture(textures[current_frame]);

            // Must also set texture rect in SFML 3
            sf::Vector2u texSize = textures[current_frame].getSize();
            sprite.setTextureRect(sf::IntRect(sf::Vector2i(0, 0), sf::Vector2i(texSize.x, texSize.y)));
        }
    }

    float dx = 0.f;
    float dy = 0.f;

    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Key::W) || sf::Keyboard::isKeyPressed(sf::Keyboard::Key::Up))
        dy -= 1.f;
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Key::S) || sf::Keyboard::isKeyPressed(sf::Keyboard::Key::Down))
        dy += 1.f;
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Key::A) || sf::Keyboard::isKeyPressed(sf::Keyboard::Key::Left))
        dx -= 1.f;
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Key::D) || sf::Keyboard::isKeyPressed(sf::Keyboard::Key::Right))
        dx += 1.f;

    if (dx != 0.f || dy != 0.f)
    {
        float magnitude = std::sqrt(dx * dx + dy * dy);
        if (magnitude != 0.f)
        {
            dx /= magnitude;
            dy /= magnitude;
        }
    }

    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Key::LShift))
        current_speed = base_speed * 0.5f;
    else
        current_speed = base_speed;

    // use float positions for dt-accurate movement
    this->position.first += dx * current_speed * dt;
    this->position.second += dy * current_speed * dt;

    // clamp in float space
    float min_x = 0.f;
    float min_y = 0.f;
    float max_x = static_cast<float>(view_w - width);
    float max_y = static_cast<float>(view_h - height);

    if (this->position.first < min_x)
        this->position.first = min_x;
    if (this->position.second < min_y)
        this->position.second = min_y;
    if (this->position.first > max_x)
        this->position.first = max_x;
    if (this->position.second > max_y)
        this->position.second = max_y;
}

void Player::draw(sf::RenderWindow &window)
{
    static bool firstDraw = true;
    if (firstDraw)
    {

        // Check sprite color
        sf::Color spriteColor = sprite.getColor();

        // Check sprite bounds
        sf::FloatRect bounds = sprite.getGlobalBounds();

        firstDraw = false;
    }

    if (textures.empty())
    {
        return;
    }

    // draw using precise float position for smooth movement
    sprite.setPosition(sf::Vector2f(this->position.first, this->position.second));

    // Get texture size and scale sprite to match width/height
    const sf::Texture &currentTexture = sprite.getTexture();
    sf::Vector2u texSize = currentTexture.getSize();

    if (texSize.x > 0 && texSize.y > 0)
    {
        float scaleX = static_cast<float>(this->width) / static_cast<float>(texSize.x);
        float scaleY = static_cast<float>(this->height) / static_cast<float>(texSize.y);
        sprite.setScale(sf::Vector2f(scaleX, scaleY));
    }

    // Make sure sprite has proper color (not transparent)
    sprite.setColor(sf::Color::White);

    window.draw(sprite);
}

std::pair<float, float> Player::get_collision()
{
    return std::pair<float, float>(this->position.first + this->width / 2.0f, this->position.second + this->height / 2.0f);
}